tests/test_mutations.py

```diff
@@ -472,45 +779,308 @@ def test_keep_mutation_parent_zero_rate(self):
+            branch_length = branch_end - branch_start
+            index = np.searchsorted(map_position, edge.left)
+            if map_position[index] > edge.left:
+                index -= 1
+            left = edge.left
+            right = 0
+            while right != edge.right:
+                right = min(edge.right, map_position[index + 1])
+                site_left = np.ceil(left) if discrete_sites else left
+                site_right = np.ceil(right) if discrete_sites else right
+                assert site_left <= site_right
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.rate[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(self.rng.poisson(mu)):
+                    position = self.rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
+                    assert edge.left <= position
+                    assert position < edge.right
+                    if position not in self.sites:
+                        site = self.add_site(
+                                        position=position,
+                                        new=True,
+                                        ancestral_state="X",
+                                        metadata=b'')
+                    time = self.rng.flat(branch_start, branch_end)
+                    self.add_mutation(
+                            id=tskit.NULL,
+                            site=site,
+                            node=edge.child,
+                            time=time,
+                            new=True,
+                            derived_state="X",
+                            metadata=b"")
+                index += 1
+                left = right
+
+    def pick_allele(self):
+        u = self.rng.flat(0, 1)
+        j = 0
+        while u > self.alphabet_distribution[j]:
+            u -= self.alphabet_distribution[j]
+            j += 1
+        return j
+
+    def transition_allele(self, j):
+        u = self.rng.flat(0, 1)
+        k = 0
+        while u > self.mutation_matrix[j][k]:
+            u -= self.mutation_matrix[j][k]
+            k += 1
+        return k
+
+    def choose_alleles(self, tree_parent, site, mutation_id):
+        if site.new:
+            ai = self.pick_allele()
+            site.ancestral_state = self.alleles[ai]
+        else:
+            ai = self.alleles.index(site.ancestral_state)
+        # sort mutations by (increasing id if both are not null,
+        #  decreasing time, increasing insertion order)
+        print([s.id for s in site.mutations])
+        site.mutations.sort(key=functools.cmp_to_key(cmp_mutation))
+        print([s.id for s in site.mutations])
+        bottom_mutation = {}
```

Comment:
Fair enough, but I opted not to because we'd have to either (a) make `bottom_mutation` a new array of length `num_nodes`, mostly empty, every time we call this function, or (b) create it outside this function and pass it in even though it's got no information in it. In the C code we do (b), but I felt like either of these two things would feel really wierd in the python code and obscure the algorithm.

Reply:
Sure, not a big deal. Pretty easy to translate between ``x not in bottom_mutation`` and ``bottom_mutation[x] == NULL``